File: apps/api/src/schemas.py
"""Pydantic schemas for validated AI outputs."""

from datetime import datetime
from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field


class JobStatus(str, Enum):
    EMPLOYED = "employed"
    SEEKING = "seeking"
    REMOTE = "remote"
    RETIRED = "retired"


class FactType(str, Enum):
    DESTINATION_PREFERENCE = "destination_preference"
    CURRENT_LOCATION = "current_location"
    FAMILY_STATUS = "family_status"
    JOB_STATUS = "job_status"
    BUDGET_RANGE = "budget_range"
    TIMELINE = "timeline"
    LANGUAGE = "language"
    VISA_REQUIREMENT = "visa_requirement"
    CUSTOM = "custom"


# Structured outputs for Pydantic AI
class DestinationPreference(BaseModel):
    """A user's destination preference."""
    country: str = Field(description="Country name")
    priority: int = Field(ge=1, le=10, description="Priority from 1-10")
    reasons: list[str] = Field(default_factory=list, description="Reasons for choosing this destination")


class FamilyCondition(BaseModel):
    """User's family situation."""
    has_partner: bool = False
    has_children: bool = False
    children_ages: list[int] = Field(default_factory=list)
    partner_nationality: Optional[str] = None
    partner_work_status: Optional[str] = None


class BudgetRange(BaseModel):
    """User's budget constraints."""
    monthly_min: Optional[int] = None
    monthly_max: Optional[int] = None
    currency: str = "USD"
    includes_housing: bool = True


class UserConditions(BaseModel):
    """Complete user conditions extracted from conversation."""
    destination_preferences: list[DestinationPreference] = Field(default_factory=list)
    family_condition: Optional[FamilyCondition] = None
    job_status: Optional[JobStatus] = None
    budget: Optional[BudgetRange] = None
    timeline: Optional[str] = None
    current_location: Optional[str] = None


class ExtractedFact(BaseModel):
    """A fact extracted from conversation."""
    type: FactType
    value: str
    confidence: float = Field(ge=0, le=1, description="Confidence score 0-1")
    requires_confirmation: bool = Field(
        default=False,
        description="Whether this fact needs user confirmation"
    )
    context: str = Field(default="", description="Context from the conversation")


class FactExtractionResult(BaseModel):
    """Result of fact extraction from conversation."""
    facts: list[ExtractedFact] = Field(default_factory=list)
    user_conditions: Optional[UserConditions] = None
    has_changes: bool = False
    summary: str = ""


class PendingConfirmation(BaseModel):
    """A pending confirmation for HITL workflow."""
    id: Optional[str] = None
    user_id: str
    fact_type: FactType
    old_value: Optional[str] = None
    new_value: str
    confidence: float
    context: str
    status: str = "pending"
    created_at: Optional[datetime] = None


class ChatMessage(BaseModel):
    """A chat message."""
    role: str = Field(pattern="^(user|assistant|system)$")
    content: str


class ChatRequest(BaseModel):
    """Request for chat endpoint."""
    messages: list[ChatMessage]
    user_id: Optional[str] = None
    session_id: Optional[str] = None


class ArticleRecommendation(BaseModel):
    """An article recommendation."""
    id: str
    title: str
    slug: str
    summary: str
    relevance_score: float
    tags: list[str] = Field(default_factory=list)


class ChatResponse(BaseModel):
    """Response from chat endpoint (non-streaming)."""
    content: str
    extracted_facts: list[ExtractedFact] = Field(default_factory=list)
    pending_confirmations: list[PendingConfirmation] = Field(default_factory=list)
    recommendations: list[ArticleRecommendation] = Field(default_factory=list)

File: packages/ai/src/schemas.ts
/**
 * Zod schemas for validated AI outputs - mirrors Python Pydantic schemas.
 */

import { z } from 'zod'

// Enums
export const JobStatusSchema = z.enum(['employed', 'seeking', 'remote', 'retired'])
export type JobStatus = z.infer<typeof JobStatusSchema>

export const FactTypeSchema = z.enum([
  'destination_preference',
  'current_location',
  'family_status',
  'job_status',
  'budget_range',
  'timeline',
  'language',
  'visa_requirement',
  'custom',
])
export type FactType = z.infer<typeof FactTypeSchema>

// Structured outputs
export const DestinationPreferenceSchema = z.object({
  country: z.string(),
  priority: z.number().min(1).max(10),
  reasons: z.array(z.string()).default([]),
})
export type DestinationPreference = z.infer<typeof DestinationPreferenceSchema>

export const FamilyConditionSchema = z.object({
  has_partner: z.boolean().default(false),
  has_children: z.boolean().default(false),
  children_ages: z.array(z.number()).default([]),
  partner_nationality: z.string().nullable().optional(),
  partner_work_status: z.string().nullable().optional(),
})
export type FamilyCondition = z.infer<typeof FamilyConditionSchema>

export const BudgetRangeSchema = z.object({
  monthly_min: z.number().nullable().optional(),
  monthly_max: z.number().nullable().optional(),
  currency: z.string().default('USD'),
  includes_housing: z.boolean().default(true),
})
export type BudgetRange = z.infer<typeof BudgetRangeSchema>

export const UserConditionsSchema = z.object({
  destination_preferences: z.array(DestinationPreferenceSchema).default([]),
  family_condition: FamilyConditionSchema.nullable().optional(),
  job_status: JobStatusSchema.nullable().optional(),
  budget: BudgetRangeSchema.nullable().optional(),
  timeline: z.string().nullable().optional(),
  current_location: z.string().nullable().optional(),
})
export type UserConditions = z.infer<typeof UserConditionsSchema>

export const ExtractedFactSchema = z.object({
  type: FactTypeSchema,
  value: z.string(),
  confidence: z.number().min(0).max(1),
  requires_confirmation: z.boolean().default(false),
  context: z.string().default(''),
})
export type ExtractedFact = z.infer<typeof ExtractedFactSchema>

export const FactExtractionResultSchema = z.object({
  facts: z.array(ExtractedFactSchema).default([]),
  user_conditions: UserConditionsSchema.nullable().optional(),
  has_changes: z.boolean().default(false),
  summary: z.string().default(''),
})
export type FactExtractionResult = z.infer<typeof FactExtractionResultSchema>

export const PendingConfirmationSchema = z.object({
  id: z.string().optional(),
  user_id: z.string(),
  fact_type: FactTypeSchema,
  old_value: z.string().nullable().optional(),
  new_value: z.string(),
  confidence: z.number(),
  context: z.string(),
  status: z.string().default('pending'),
  created_at: z.string().optional(),
})
export type PendingConfirmation = z.infer<typeof PendingConfirmationSchema>

export const ChatMessageSchema = z.object({
  role: z.enum(['user', 'assistant', 'system']),
  content: z.string(),
})
export type ChatMessage = z.infer<typeof ChatMessageSchema>

export const ArticleRecommendationSchema = z.object({
  id: z.string(),
  title: z.string(),
  slug: z.string(),
  summary: z.string(),
  relevance_score: z.number(),
  tags: z.array(z.string()).default([]),
})
export type ArticleRecommendation = z.infer<typeof ArticleRecommendationSchema>

export const ChatResponseSchema = z.object({
  content: z.string(),
  extracted_facts: z.array(ExtractedFactSchema).default([]),
  pending_confirmations: z.array(PendingConfirmationSchema).default([]),
  recommendations: z.array(ArticleRecommendationSchema).default([]),
})
export type ChatResponse = z.infer<typeof ChatResponseSchema>
